raiden/tasks.py

```diff
@@ -134,6 +135,78 @@ def stop_async(self):
         self.stop_event.set(True)
 
 
+class HealthcheckTask(Task):
+    """ Task for checking if all of our open channels are healthy """
+
+    def __init__(
+            self,
+            raiden,
+            send_ping_time,
+            max_unresponsive_time,
+            sleep_time=DEFAULT_HEALTHCHECK_POLL_TIMEOUT):
+        """ Initialize a HealthcheckTask that will monitor open channels for
+             responsiveness.
+
+             :param raiden RaidenService: The Raiden service which will give us
+                                          access to the protocol object and to
+                                          the asset manager
+             :param int sleep_time: Time in seconds between each healthcheck task
+             :param int send_ping_time: Time in seconds after not having received
+                                        a message from an address at which to send
+                                        a Ping.
+             :param int max_unresponsive_time: Time in seconds after not having received
+                                               a message from an address at which it
+                                               should be deleted.
+         """
+        super(HealthcheckTask, self).__init__()
+
+        self.protocol = raiden.protocol
+        self.raiden = raiden
+
+        self.stop_event = AsyncResult()
+        self.sleep_time = sleep_time
+        self.send_ping_time = send_ping_time
+        self.max_unresponsive_time = max_unresponsive_time
+
+    def _run(self):  # pylint: disable=method-hidden
+        stop = None
+        while stop is None:
+            keys_to_remove = []
+            for key, queue in self.protocol.address_queue.iteritems():
```

Comment:
Instead of adding a new task, we could add the ping logic in the `_send_queued_message`:

- Send ping to check if partner is online
    - if ping unconfirmed -> node is unavailable
        - wait for partner to come online (if any message is received from partner)
- loop:
  - if there is a message in the queue
    - until message confirmed, have retries, or partner is back online
       - send message
       - wait for confirmation or timeout
       - if max_retries == 0 node is unavailable (the queue message worked and the Ping)
          - wait for partner to come back online (available if any message is received from partner)
          - restart max_retries counter
        - elseif unconfirmed, decrement max_retries
  - if no message in the queue and `send-ping-time` passed
    - send Ping and wait for confirmation
    - if ping unconfirmed -> node is unavailable
        - wait for partner to come back online (available if any message is received from partner)

This would make one task per channel, we eliminate synchronizations issues for messages that have `max_retries`, and we can start these queues in a randomization loop as suggested by @konradkonrad